core/kotlinx-coroutines-debug/src/debug/internal/DebugProbesImpl.kt

```diff
@@ -0,0 +1,282 @@
+                val observedStackTrace = value.lastObservedStackTrace()
+                if (observedStackTrace.isEmpty()) {
+                    append("\n\tat ${artificialFrame(ARTIFICIAL_FRAME_MESSAGE)}")
+                    printStackTrace(value.creationStackTrace)
+                } else {
+                    printStackTrace(value.lastObservedStackTrace())
+                }
+            }
+        }
+
+        // Move it out of synchronization?
+        out.println(resultingString)
+    }
+
+    private fun StringBuilder.printStackTrace(frames: List<StackTraceElement>) {
+        frames.forEach { frame ->
+            append("\n\tat $frame")
+        }
+    }
+
+    @Synchronized
+    internal fun probeCoroutineResumed(frame: Continuation<*>) = updateState(frame, State.RUNNING)
+
+    @Synchronized
+    internal fun probeCoroutineSuspended(frame: Continuation<*>) = updateState(frame, State.SUSPENDED)
+
+    private fun updateState(frame: Continuation<*>, state: State) {
+        if (!isInstalled) {
+            return
+        }
+
+        // Find ArtificialStackFrame of the coroutine
+        val owner = frame.owner()
+        val coroutineState = capturedCoroutines[owner]
+        if (coroutineState == null) {
+            warn(frame, state)
+            return
+        }
+
+        coroutineState.updateState(state, frame)
+    }
+
+    private fun Continuation<*>.owner(): ArtificialStackFrame<*>? {
+        var frame = this as? CoroutineStackFrame ?: return null
+        while (true) {
+            if (frame is ArtificialStackFrame<*>) return frame
+            val completion = frame.callerFrame ?: return null
+            frame = completion
+        }
+    }
+
+    @Synchronized
+    internal fun <T> probeCoroutineCreated(completion: Continuation<T>): Continuation<T> {
+        if (!isInstalled) {
+            return completion
+        }
+
+        /*
+         * Here we replace completion with a sequence of CoroutineStackFrame objects
+         * which represents creation stacktrace, thus making stacktrace recovery mechanism
+         * even more verbose (it will attach coroutine creation stacktrace to all exceptions),
+         * and then using this artificial frame as an identifier of coroutineSuspended/resumed calls.
+         */
+        val stacktrace = sanitizedStackTrace(Exception())
+        val frames = ArrayList<CoroutineStackFrame?>(stacktrace.size)
+        for ((index, frame) in stacktrace.reversed().withIndex()) {
+            frames += object : CoroutineStackFrame {
+                override val callerFrame: CoroutineStackFrame?
+                    get() = if (index == 0) null else frames[index - 1]
+
+                override fun getStackTraceElement(): StackTraceElement = frame
+            }
+        }
+
+        val result = ArtificialStackFrame(completion, frames.last()!!)
+        capturedCoroutines[result] = CoroutineState(completion, stacktrace.slice(1 until stacktrace.size), ++sequenceNumber)
```

Comment:
It seems inefficient to store an _array_ of stack-traces in `CoroutineState` when you already have created a linked list of frames that you have to retain in memory anyway.